src/transformers/trainer.py

```diff
@@ -811,8 +836,9 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
                             # AMP: gradients need unscaling
                             self.scaler.unscale_(self.optimizer)
 
-                        if hasattr(self.optimizer, "clip_grad_norm"):
+                        if hasattr(self.optimizer, "clip_grad_norm") and not self.args.deepspeed:
                             # Some optimizers (like the sharded optimizer) have a specific way to do gradient clipping
+                            # deepspeed has clip_grad_norm aliased to torch.nn.utils.clip_grad_norm_
```

Comment:
Is this complexity really necessary? E.g. is `torch.nn.utils.clip_grad_norm_` doing something different than `self.optimizer.clip_grad_norm`?

Reply:
yes, 

1. the 2 functions have a different argument signature - that's why I had to add this. or have I missed something?

# pytorch
torch.nn.utils.clip_grad_norm_(parameters: Union[torch.Tensor, Iterable[torch.Tensor]], max_norm: float, norm_type: float = 2.0) → torch.Tensor

# fairscale
def clip_grad_norm(self, max_norm: Union[float, int], norm_type: Union[float, int] = 2.0) -> torch.Tensor

2. deepspeed provides `clip_grad_norm` just like fairscale does, but it aliases it to `torch.nn.utils.clip_grad_norm_`

Does this make sense?

Reply:
I'd say this perhaps is a bug in DS? Why create a feature method if you are not handling it as such. Should I ask about it? Perhaps it's some old-times left-over and they could just remove it altogether (but it'd take a deprecation cycle).

Reply:
If it's not usable, I'd say it's a bug in DS. No problem with the test then, I was just wondering why the extra stuff here.

Reply:
Yes, please see https://github.com/microsoft/DeepSpeed/issues/611 - they are going to review this and hopefully clean it up.

Reply:
But I don't think we can do anything about it at the moment